### PR TITLE
[FXML-1981] Remove QDQ nodes between foldable concats

### PR DESCRIPTION
**Pull request by gargaroff**

### CODE REVIEW

**Comment by gargaroff** [comment]
After discussing with @mgehre-amd and @cmcgirr-amd I will move the concat fold into LLVM, so we can upstream it eventually

File: include/xten/Transform/XTenCanonicalizePass.h [hunk 1]
```diff
@@ -11,7 +11,7 @@ namespace xilinx::xten {
 #include "xten/Transform/Passes.h.inc"
 
 /// Populate \p patterns with XTen concat folding patterns.
-void populateXTenFoldConcatPatterns(mlir::RewritePatternSet &patterns, mlir::PatternBenefit benefit = 1);
+void populateXTenQDQConcatPatterns(mlir::RewritePatternSet &patterns);
```
Comment (mgehre-amd) [inline include/xten/Transform/XTenCanonicalizePass.h]: Can we add a big warning here and in the source code of the pattern that this is destructive and changes the semantics of the IR?

File: lib/Transform/XTenQDQConcat.cpp [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+namespace {
+struct XTenQDQConcat : public OpRewritePattern<DequantizeOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(DequantizeOp op,
+                                PatternRewriter &rewriter) const override {
+    // Match concat->QDQ->concat and remove QDQ, if concats would be foldable.
+    // Try to match an incoming concat
+    Value val;
+    if (!matchPattern(op.getInput(), m_Op<QuantizeOp>(matchers::m_Any(&val)))) {
```
Comment (mgehre-amd) [inline lib/Transform/XTenQDQConcat.cpp]: I think you could write the concat-Quantize-dequantize-concat as a single call to `matchPattern`, but I have no opinion on whether that would be more readable (it would look more like PDLL)

File: lib/Transform/XTenQDQConcat.cpp [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+    auto producer = dyn_cast_or_null<ConcatOp>(val.getDefiningOp());
+    if (producer == nullptr) {
+      return rewriter.notifyMatchFailure(
+          op, "QDQ input not produced by TOSA concat.");
+    }
+
+    uint64_t axis = producer.getAxis();
+
+    bool modified = false;
+    for (Operation *user : op->getUsers()) {
```
Comment (mgehre-amd) [inline lib/Transform/XTenQDQConcat.cpp]: Should we update some users if we cannot update all users? It seems that in that case the concats cannot be folded.

Then we can we express the check via an `llvm::all_of` and the update via a single `rewriter.replaceOp(op, producer.getOutput()`

File: lib/Transform/XTenQDQConcat.cpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+  LogicalResult matchAndRewrite(DequantizeOp op,
+                                PatternRewriter &rewriter) const override {
+    // Match concat->QDQ->concat and remove QDQ, if concats would be foldable.
+    // Try to match an incoming concat
+    Value val;
+    if (!matchPattern(op.getInput(), m_Op<QuantizeOp>(matchers::m_Any(&val)))) {
+      return rewriter.notifyMatchFailure(
+          op, "Dequantize input not produced by Quantize.");
+    }
```
Comment (mgehre-amd) [inline lib/Transform/XTenQDQConcat.cpp]: To be a bit less destructive, should we check that quantize and dequantize have the same scale?

**Review by mgehre-amd — CHANGES_REQUESTED** [review]

File: lib/Transform/XTenQDQConcat.cpp (hunk 2 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: I tried to do that but didn't get it to work. It simply would not compile. This is what I tried to do:
```c++
ConcatOp concat;
if (!matchPattern(op.getInput(), m_Op<QuantizeOp>(m_Op<ConcatOp>(&concat)))
```

File: lib/Transform/XTenQDQConcat.cpp (hunk 3 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: Since we only update the user and do not otherwise touch the matched quantize node, I don't see a reason why we shouldn't update just some of them to enable at least some folding

File: lib/Transform/XTenQDQConcat.cpp (hunk 4 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: That's actually a good check. Will do

File: lib/Transform/XTenQDQConcat.cpp (hunk 2 above)
Comment (mgehre-amd) [reply lib/Transform/XTenQDQConcat.cpp]: @cmcgirr-amd seem to have figured that out here: https://github.com/ljfitz/mlir-xten/pull/66/files#diff-b9665df2774af4535d3f49fdec1a326cbcabec850dfb7e3d47ef01117a7780acR163

Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: Unfortunately this wouldn't work, as it is matching something slightly different. I need to match the concat itself. The linked pattern however is matching the inputs to the mul op. It seems like there are only binding matchers for any values and constants, but not for concrete ops.

Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: With the new shift check this became obsolete, as I cannot use `matchPattern` anymore.

File: include/xten/Transform/XTenCanonicalizePass.h (hunk 1 above)
Comment (gargaroff) [reply include/xten/Transform/XTenCanonicalizePass.h]: I split the patterns into destructive and non-destructive patterns and have hidden the destructive ones behind a command line option.

**Review by mgehre-amd — APPROVED** [review]

File: include/xten/Transform/XTenCanonicalizePass.h [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#pragma once
```
Comment (cmcgirr-amd) [inline include/xten/Transform/XTenCanonicalizePass.h]: nit: all the other headers use the old style header guard of `#ifdefs`

File: lib/Transform/XTenQDQConcat.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+#include <mlir/Dialect/Tosa/IR/TosaOps.h>
+#include <mlir/IR/Matchers.h>
+#include <mlir/IR/PatternMatch.h>
+#include <mlir/Support/LogicalResult.h>
+
+#include <llvm/ADT/SmallVector.h>
+
+using namespace mlir;
+using namespace mlir::tosa;
+using namespace amd::xten_nn;
```
Comment (cmcgirr-amd) [inline lib/Transform/XTenQDQConcat.cpp]: Would it make sense to also use this namespace for these passes? The operations we work on are XTenNN dialect operations rather than XTen.

File: include/xten/Transform/Passes.td [hunk 7]
```diff
@@ -50,4 +50,20 @@ def XTenMinimizeLiveTensors : Pass<"xten-minimize-live", "func::FuncOp"> {
   ];
 }
 
+def XTenCanonicalizePass : Pass<"xten-canonicalize"> {
+  let summary = "Perform XTen specific canonicalizations on the IR";
+  let description = [{
+    Pass that performs XTen specific foldings & canonicalizations to bring the
```
Comment (cmcgirr-amd) [inline include/xten/Transform/Passes.td]:
```suggestion
    Pass that performs XTenNN specific foldings & canonicalizations to bring the
```
I know its confusing but we have a dialect named XTen and another named XTenNN with our QDQ ops defined.

**Comment by cmcgirr-amd** [comment]
The new files are missing AMD 2023 copyright notices.  I think we should add them.

File: include/xten/Transform/XTenCanonicalizePass.h [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+#include <mlir/IR/MLIRContext.h>
```
Comment (cmcgirr-amd) [inline include/xten/Transform/XTenCanonicalizePass.h]: unused header?

File: lib/Transform/XTenQDQConcat.cpp (hunk 3 above)
Comment (cmcgirr-amd) [reply lib/Transform/XTenQDQConcat.cpp]: But a partial rewrite of the QDQ nodes does not enable Concat folding later on in the case where only a part of the users match. So we destroy semantics with no benefit.

File: test/Transform/XTenCanonicalizePass/remove_qdq_between_concats.mlir [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+  %4 = "tosa.concat"(%2, %2) {axis = 0} : (tensor<1x2x7x7xf32>, tensor<1x2x7x7xf32>) -> tensor<2x2x7x7xf32> // This is not foldable
+  return %3, %4 : tensor<1x4x7x7xf32>, tensor<2x2x7x7xf32>
+}
+
+// CHECK-LABEL:   func.func @partially_foldable_user_concats(
+// CHECK-SAME:                                               %[[VAL_0:.*]]: tensor<1x1x7x7xf32>) -> (tensor<1x4x7x7xf32>, tensor<2x2x7x7xf32>) {
+// CHECK:           %[[VAL_1:.*]] = "tosa.concat"(%[[VAL_0]], %[[VAL_0]]) {axis = 1 : i64} : (tensor<1x1x7x7xf32>, tensor<1x1x7x7xf32>) -> tensor<1x2x7x7xf32>
+// CHECK:           %[[VAL_2:.*]] = xten_nn.quantize(%[[VAL_1]] : tensor<1x2x7x7xf32>) {shift = -3 : si32} -> tensor<1x2x7x7xsi8>
+// CHECK:           %[[VAL_3:.*]] = xten_nn.dequantize(%[[VAL_2]] : tensor<1x2x7x7xsi8>) {shift = -3 : si32} -> tensor<1x2x7x7xf32>
+// CHECK:           %[[VAL_4:.*]] = "tosa.concat"(%[[VAL_1]], %[[VAL_1]]) {axis = 1 : i64} : (tensor<1x2x7x7xf32>, tensor<1x2x7x7xf32>) -> tensor<1x4x7x7xf32>
```
Comment (cmcgirr-amd) [inline test/Transform/XTenCanonicalizePass/remove_qdq_between_concats.mlir]: This test case does not enable concat folding later. Is there a case where multiple uses of the DQ node and partial folding leads to concat folding later?

**Comment by cmcgirr-amd** [comment]
I think we should also move the pass to `xten/Dialect/XTenNN/Transforms`. I know super annoying but it would make sense.

File: test/Transform/XTenCanonicalizePass/remove_qdq_between_concats.mlir (hunk 9 above)
Comment (gargaroff) [reply test/Transform/XTenCanonicalizePass/remove_qdq_between_concats.mlir]: It does enable folding though, at least for one of the two concats:

Before:
```mlir
// Both concats look like this:
%0 = concat
%1 = quantize
%2 = dequantize
%3 = concat
```

After
```mlir
// One concat looks like this:
%0 = concat
%1 = quantize
%2 = dequantize
%3 = concat

// The other now looks like this and can be folded:
%0 = concat
%2 = concat
```

File: lib/Transform/XTenQDQConcat.cpp (hunk 3 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: This is not what is happening here. See the response to other comment on the test case

File: lib/Transform/XTenQDQConcat.cpp (hunk 6 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: Do you mean the `populateXTenQDQConcatPatterns` and the `createXTenCanonicalizePass`?

File: include/xten/Transform/XTenCanonicalizePass.h (hunk 8 above)
Comment (gargaroff) [reply include/xten/Transform/XTenCanonicalizePass.h]: Unused indeed, thanks!

File: include/xten/Transform/XTenCanonicalizePass.h (hunk 5 above)
Comment (gargaroff) [reply include/xten/Transform/XTenCanonicalizePass.h]: Consider it a lazy, on-demand migration to the new style 😛

File: lib/Transform/XTenQDQConcat.cpp (hunk 6 above)
Comment (cmcgirr-amd) [reply lib/Transform/XTenQDQConcat.cpp]: Yes I mean everything that is new here, we have our XTenNN under the new amd namespace. Which aligns better with the current state.

Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: Sounds good. I'll move it!

**Comment by gargaroff** [comment]
> I think we should also move the pass to xten/Dialect/XTenNN/Transforms. I know super annoying but it would make sense.

Totally agree and not too annoying. Gives me an opportunity to implement passes the right way :)

File: include/xten/Transform/XTenCanonicalizePass.h (hunk 5 above)
Comment (gargaroff) [reply include/xten/Transform/XTenCanonicalizePass.h]: Since I will be moving this pass to XTenNN, I will leave this as-is, since the passes (or rather, pass; there's only one) use this style already

File: lib/Transform/XTenQDQConcat.cpp [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    if (modified)
+      return success();
+
+    return rewriter.notifyMatchFailure(
+        op, "No foldable concats found around QDQ node.");
+  }
+};
+} // namespace
+
+void xilinx::xten::populateXTenQDQConcatPatterns(RewritePatternSet &patterns) {
```
Comment (ttjost) [inline lib/Transform/XTenQDQConcat.cpp]: Will we have more than one pattern here? Otherwise, I would argue we don't need to have `populateXTenQDQConcatPatterns` but just the single pattern added in `populateXTenDestructiveCanonicalizePatterns`.

**Review by cmcgirr-amd — APPROVED** [review]
Thanks you! Looks good :)

File: lib/Transform/XTenQDQConcat.cpp (hunk 10 above)
Comment (cmcgirr-amd) [reply lib/Transform/XTenQDQConcat.cpp]: We will probably also have wrapping of maxpool/reshape with QDQs here as well in the `populateXTenQDQConcatPatterns`.

**Review by ttjost — APPROVED** [review]
Just one comment about whether we would have multiple patterns in `populateXTenQDQConcatPatterns`. Thanks!

File: lib/Transform/XTenQDQConcat.cpp (hunk 10 above)
Comment (gargaroff) [reply lib/Transform/XTenQDQConcat.cpp]: The idea is two fold:
1. You can add more QDQConcat patterns in the future
2. It allows you to use these patterns without having to use all of the destructive patterns

**Comment by gargaroff** [comment]
@ttjost I noticed that you did your review before I refactored it to be part of XTenNN. The implementation didn't change, just moving boilerplate

File: lib/Dialect/XTenNN/Transforms/QDQConcat.cpp [hunk 11]
```diff
@@ -1,19 +1,19 @@
 #include "xten/Dialect/XTenNN/IR/XTenNNOps.h"
-#include "xten/Transform/Passes.h"
+#include "xten/Dialect/XTenNN/Transforms/CanonicalizePass.h"
 
-#include <mlir/Dialect/Tosa/IR/TosaOps.h>
-#include <mlir/IR/Matchers.h>
-#include <mlir/IR/PatternMatch.h>
-#include <mlir/Support/LogicalResult.h>
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
```
Comment (ttjost) [inline lib/Dialect/XTenNN/Transforms/QDQConcat.cpp:18]: Any reason why this change here?

Comment (gargaroff) [reply lib/Dialect/XTenNN/Transforms/QDQConcat.cpp:18]: The other pass in XtenNN is using this style for llvm/mlir includes, so I changed this to use the same style
